torchdynamo/utils.py

```diff
@@ -416,6 +416,64 @@ def check_constant_args(args, kwargs):
     return all(x.is_python_constant() for x in itertools.chain(args, kwargs.values()))
 
 
+def check_unspec_python_args(args, kwargs):
+    from .variables.constant import ConstantVariable
+    from .variables.tensor import FakeItemVariable
+    from .variables.tensor import UnspecializedPythonVariable
+
+    return (
+        all(
+            isinstance(
+                i,
+                (
+                    UnspecializedPythonVariable,
+                    ConstantVariable,
+                ),
+            )
+            for i in itertools.chain(args, kwargs.values())
+        )
+        and any(
+            isinstance(x, UnspecializedPythonVariable)
+            for x in itertools.chain(args, kwargs.values())
+        )
+        and not any(
+            isinstance(x, FakeItemVariable)
+            for x in itertools.chain(args, kwargs.values())
+        )
+    )
+
+
+def specialize_args_kwargs(tx, args, kwargs):
+    from .variables.tensor import UnspecializedNumpyVariable
+    from .variables.tensor import UnspecializedPythonVariable
+
+    specialized_args = []
+    specialized_kwargs = {}
+    for x in args:
+        if isinstance(
+            x,
+            (
+                UnspecializedNumpyVariable,
+                UnspecializedPythonVariable,
+            ),
+        ):
+            specialized_args.append(x.convert_to_constant(tx))
+        else:
+            specialized_args.append(x)
```

Comment:
I am a little confused by this - if args are unspecialized, we record them as constants, but otherwise we record them as is? If this is to provide safety for upstream of places where we check for `isinstance(foo, variables.ConstantVariable)`, we may be better off making a helper function like `is_constant_convertible(foo)` and  `to_constant(foo)`?

Reply:
This is a fair point, but I had addressed in this [comment](https://github.com/pytorch/torchdynamo/pull/714#discussion_r938293298). The short answer is: 1, it's rare case,  never occurs for all torchbench models, so I keep it as is to make this PR not too big to review; 2, I'll have another PR which adding a global data structure to store all local variables <-> source <-> spec/unspec, and make a helper function like you pointed out.

Reply:
We don't need it to be global, I think, and this is actually a slightly different question from mapping altogether.

The question is, why not make some helper methods that make it so that we don't need to do all this conversion ahead of time? Why not just let places that take `ConstVariable` also take `UnspecializedNumpyVariable` and `UnspecializedPythonVariable`, and then just give them a `to_const_variable() -> ConstVariable` type?

Reply:
Sorry I can't catch your point, can you elaborate more or do you have example?

Reply:
Sure thing! The pattern we usually follow is not `x.convert_to_foo()` its `x.as_foo()` where `as_foo` is not destructive - it does not change the original object, but produces a new one. What's cool about that is we can have

```
class ConstantVariable :
def as_constant(tx) -> ConstantVariable:
     # identity
    return self
    
...
```

and 
```
UnspecializedVariable
def as_constant(tx) -> ConstantVariable:
    # your existing logic for making a ConstantVariable
    return ConstantVariable(...)
...
```

What's cool about this is you now dont need to do any iteration or pre-processing to turn out `ConstantVariable` from your `Unspecialized` objects, you can just pass around `Unspecialized` everywhere, and as long as you call `.as_constant()` right before you use them, it will be safe, and easier to read, and easier to extend if we get another type that needs to support `.as_constant()`

Reply:
Fair point, it can remove the pre-processing and always call ```as_constant```.

Reply:
Wonderful, thank you sir, sorry it took me so long to communicate what I meant clearly.